### ISSUE TITLE
Assertion failed: (I != M.SubmoduleRemap.end() && "Invalid index into submodule index remap")

### ISSUE THREAD

**Issue by rmaz** [clangd/clangd#198]
When running clangd on files that import headers from their own modules, the following assertion is hit:

```
/Users/richardh/Code/dummy-toolchain/bin/clang-tool -I. -fmodules -fmodule-name=MyModule /Users/richardh/Code/moduletestcase/src.m -fsyntax-only -resource-dir=/Users/richardh/Code/dummy-toolchain/lib/clang/7.0.0
Assertion failed: (I != M.SubmoduleRemap.end() && "Invalid index into submodule index remap"), function getGlobalSubmoduleID, file /Users/buildnode/jenkins/workspace/oss-swift-package-osx/llvm/tools/clang/lib/Serialization/ASTReader.cpp, line 8933.
```

Compiling with the same flags produces no error. I've pushed a test case repo here:

https://github.com/rmaz/clangdmoduletestcase

**Comment by rmaz** [comment]
This also manifests as:

```
{
	"resource": "blah",
	"owner": "_generated_diagnostic_collection_name_#0",
	"code": "fe_pch_malformed",
	"severity": 8,
	"message": "Malformed or corrupted AST file: 'submodule ID out of range in AST file'",
	"source": "clang",
	"startLineNumber": 1,
	"startColumn": 1,
	"endLineNumber": 1,
	"endColumn": 1
}
```

**Comment by sam-mccall** [comment]
Thanks for reducing this!

As you may have guessed, clangd doesn't understand modules well yet.
My guess for this particular error: clangd makes heavy use of the preamble optimization, which uses the same underlying machinery as modules, and the two are being confused somehow.

@hyp @koja may have ideas here.

**Comment by sam-mccall** [comment]
(This reproduces on linux, but AIUI implicit modules are mostly deployed on mac, and explicit modules are mostly deployed nowhere at this point)

**Comment by hyp** [comment]
I'll try to check what's up.

**Comment by hyp** [comment]
@sam-mccall It looks like the preambles in Clangd are a bit of a hack, and assume they can only include #include directives? It looks like importing modules with preambles is unsupported.

**Comment by hyp** [comment]
@rmaz Are you intentionally using `-fmodules` without `-fimplicit-modules`? Since you don't have `-fimplicit-modules`, you would need to pass in the pre-built module file as an argument to the compiler as well.

**Comment by hyp** [comment]
In any case, for now I added a test case for `-fimplicit-modules -fmodules` to Clangd that's based on the repo you linked to: https://reviews.llvm.org/D71586.

**Comment by rmaz** [comment]
I understood implicit modules were the default and could only be disabled with `-fno-implicit-modules`? That seems to be the case from reading https://clang.llvm.org/docs/Modules.html, and is not set by Xcode projects either, just `-fmodules`

**Comment by gislan** [comment]
This was fixed a while ago in e2d61ae5733316a14783b36c84b8e7681b0e3d59 (https://reviews.llvm.org/D85532)
